File: 남은/프로그래머스/섹션8/타겟넘버.js
// https://programmers.co.kr/learn/courses/30/lessons/43165

const numbers = [1, 1, 1, 1, 1],
  target = 3;

function solution(numbers, target) {
  var answer = 0;

  function DFS(idx, sum) {
    if (idx === numbers.length) {
      if (sum === target) {
        // if (sum === target && idx === numbers.length) {
        answer += 1;
      }
    } else {
      //   console.log({ idx });
      DFS(idx + 1, sum + numbers[idx]);
      DFS(idx + 1, sum - numbers[idx]);
    }
  }

  DFS(0, 0);
  return answer;
}

console.log(solution(numbers, target));

//////////////////////
// let cnt = 0;

// function DFS(index, sum, len, numbers, target) {
//   if (index === len && sum === target) {
//     cnt += 1;
//   } else {
//     DFS(index + 1, sum + numbers[index], len, numbers, target);
//     DFS(index + 1, sum - numbers[index], len, numbers, target);
//   }
// }

// function solution(numbers, target) {
//   var answer = 0;
//   let len = numbers.length;

//   DFS(0, 0, len, numbers, target);

//   answer = cnt;
//   return answer;
// }

File: 영훈/Programmers/C++/DFS, BFS/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;
int _target;
vector<int> _numbers;

void dfs(int l, int sum) {
    if (l == _numbers.size()) {
        if (sum == _target) answer++;
    } else {
        dfs(l + 1, sum - _numbers[l]);
        dfs(l + 1, sum + _numbers[l]);
    }
}

// void dfs(int l, int sum, vector<int> numbers) {
//     if (l == numbers.size()) {
//         if (sum == _target) answer++;
//     } else {
//         dfs(l + 1, sum - numbers[l], numbers);
//         dfs(l + 1, sum + numbers[l], numbers);
//     }
// }

int solution(vector<int> numbers, int target) {
    _numbers = numbers;
    _target = target;
    
    // dfs(0, 0, numbers);
    dfs(0, 0);

    return answer;
}